auth/src/main/java/org/aerogear/auth/impl/UserPrincipalImpl.java

```diff
@@ -29,7 +27,7 @@
     /**
      * Roles associated with this principal.
      */
-    private final Map<String, IRole> roles;
+    private final List<IRole> roles;
```

Comment:
I think saving the roles inside a map would be more efficient for the 'hasRole' methods.

Reply:
The issue with this is that rolenames are not unique.  Role's can have the same name at the realm and client level so I'm not sure this would work?

Reply:
A Key object with `role name` and `role type` should be implemented.
Then the map should become `private final Map <RoleKey, IRole> roles`.

In the `RoleKey` object you should override the `equals` method so that both role name and role type gets compared. Then it should work.